pkg/clients/secrets/secrets.go

```diff
@@ -0,0 +1,138 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package secrets
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/crossplane-contrib/provider-github/apis/secrets/v1alpha1"
+	ghclient "github.com/crossplane-contrib/provider-github/pkg/clients"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/google/go-github/v33/github"
+	"golang.org/x/crypto/nacl/box"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Service defines the Secrets operations
+type Service interface {
+	GetRepoPublicKey(ctx context.Context, owner, repo string) (*github.PublicKey, *github.Response, error)
+	GetRepoSecret(ctx context.Context, owner, repo, name string) (*github.Secret, *github.Response, error)
+	CreateOrUpdateRepoSecret(ctx context.Context, owner, repo string, eSecret *github.EncryptedSecret) (*github.Response, error)
+	DeleteRepoSecret(ctx context.Context, owner, repo, name string) (*github.Response, error)
+}
+
+// NewService creates a new Service based on the *github.Client
+// returned by the NewClient SDK method.
+func NewService(token string) *Service {
+	c := ghclient.NewClient(token)
+	r := Service(c.Actions)
+	return &r
+}
+
+// CreateOrUpdateSec create or update repository secret in GitHub
+func CreateOrUpdateSec(ctx context.Context, cr *v1alpha1.SecretsParameters, name string, client client.Client, gh Service) (string, error) {
+	encryptedSecret, hash, err := callEncryptSecret(ctx, client, cr, name, gh)
+	if err != nil {
+		return "", err
+	}
+
+	if _, err := gh.CreateOrUpdateRepoSecret(ctx, cr.Owner, cr.Repository, encryptedSecret); err != nil {
+		return "", err
+	}
+
+	return hash, nil
+}
+
+// IsUpToDate check if encrypted value is up to date
+func IsUpToDate(ctx context.Context, client client.Client, p *v1alpha1.SecretsParameters, o *v1alpha1.SecretsObservation, name string, gh Service) (bool, error) {
+	var hash string
+	sec, _, err := gh.GetRepoSecret(ctx, p.Owner, p.Repository, name)
+	if err != nil {
+		return false, err
+	}
+
+	ref := xpv1.CommonCredentialSelectors{SecretRef: p.Value}
+	val, err := resource.ExtractSecret(ctx, client, ref)
+	if err != nil {
+		return false, err
+	}
+
+	if hash = generateHash(string(val)); len(val) == 0 {
+		hash = generateHash("test")
+	}
+
+	if hash != *o.EncryptValue {
+		return false, nil
+	}
+
+	if sec.UpdatedAt.String() != *o.LastUpdate {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// callEncryptedSecret setup encrypted secret and generates hash to store
+func callEncryptSecret(ctx context.Context, client client.Client, cr *v1alpha1.SecretsParameters, name string, gh Service) (*github.EncryptedSecret, string, error) {
```

Comment:
I think it is better to rename this funtion to `setupEncryptedSecret`, because of the above description